### PR TITLE
Add registry to discover driver in different platforms

### PR DESCRIPTION
**Pull request by anfernee**
Different platform has differnet list of supported drivers. The
registry contains the correct list of drivers that are supported. In
future we could add commands like `minikube list-drivers`

### CODE REVIEW

**Comment by anfernee** [comment]
I found the machine package has very similar stuff. I think it's possible to merge them. I will update this PR. but @r2d4 feel free to review this.

**Comment by afbjorklund** [comment]
Probably related to #2527 ?

**Comment by anfernee** [comment]
@afbjorklund kind of, but not just that. the intention is to remove the hardcoded `switch` statement in different packages to manage hardcoded driver names, and use a centralized component to deal with drivers (marshal/unmarshal, initialization, etc). Another thing is drivers are platform dependent, but those `switch`es are not.

**Comment by gbraad** [comment]
@praveenkumar WDYT?

**Comment by dlorenc** [comment]
@minikube-bot ok to test

**Comment by anfernee** [comment]
Guys, I moved all driver logic together and register them in registry, so it's easier to add new driver support in minikube (potentally minishift and other tools). simply register with a name, and couple of functions. 

Specifically: 
- what platform it supports?
- It's built-in or not
- a function to translate from minikube config to driver config
- a function to create a driver (only for builtin).

@dlorenc  for use cases, the top of my mind is #2527 all supported drivers were hardcoded. With this change, we only need to call `registry.ListDrivers()`. It will tell you all supported drivers on your platform, whether it's builtin or not. It can potentially tell you where to get the plugin, and even help download and install the plugin for you. Just adding those to `DriverDef` and it's all possible.

**Comment by gbraad** [comment]
@minishift/minishift-dev Please, have a look

**Comment by anfernee** [comment]
@praveenkumar I think you can import individual packages into your vendor folder without importing the whole repo. There are tools that can do that. 

You probably need those packages:
- k8s.io/minikube/pkg/minikube/registry
- k8s.io/minikube/pkg/minikube/config
- k8s.io/minikube/pkg/minikube/drivers/...

The config package might not make sense for minishift, because you probably have different `MachineConfig`. Anyway, I agree the `minikube` in package is strange. I think it would be better if it is more neutral.

**Comment by anfernee** [comment]
@praveenkumar Just try doing this using `dep`:

``` shell
$ cat main.go 
package main

import (
	"fmt"

	"k8s.io/kubernetes/pkg/util/slice"
)

func main() {
	fmt.Println(slice.SortStrings([]string{"foo", "bar"}))
}
$ dep init 
  Using ^1.9.3 as constraint for direct dep k8s.io/kubernetes
  Locking in v1.9.3 (d283541) for direct dep k8s.io/kubernetes
  Locking in master (cced8e6) for transitive dep k8s.io/apimachinery
```

It will not copy everything into vendor.
 
In terms of minishift's machine config, I think there are several ways to make it more generic. One of them is to take a viper like interface, instead of `confg.MachineConfig`, so that use `GetInt`, `GetBool` to retrieve machine's config information.

**Comment by praveenkumar** [comment]
> It will not copy everything into vendor.

@anfernee Looks like it does most of the things.

```
$ cat Gopkg.lock 
# This file is autogenerated, do not edit; changes may be undone by the next 'dep ensure'.
[[projects]]
  branch = "master"
  name = "k8s.io/apimachinery"
  packages = ["pkg/util/rand"]
  revision = "cced8e64b6ca92a8b6afcbfea3353ca016694a45"

[[projects]]
  name = "k8s.io/kubernetes"
  packages = ["pkg/util/slice"]
  revision = "d2835416544f298c919e2ead3be3d0864b52323b"
  version = "v1.9.3"

[solve-meta]
  analyzer-name = "dep"
  analyzer-version = 1
  inputs-digest = "690bb1977ae5ee1ddd17d183f087cb11f8d54154bda0d0012f76babae7fc4d66"
  solver-name = "gps-cdcl"
  solver-version = 1

$ ls vendor/k8s.io/
apimachinery  kubernetes

$ ls vendor/k8s.io/kubernetes/
BUILD.bazel       CONTRIBUTING.md  Makefile                  OWNERS_ALIASES  Vagrantfile  build    code-of-conduct.md  hack         pkg      test
CHANGELOG-1.9.md  Godeps           Makefile.generated_files  README.md       WORKSPACE    cluster  docs                labels.yaml  plugin   third_party
CHANGELOG.md      LICENSE          OWNERS                    SUPPORT.md      api          cmd      examples            logo         staging  translations

$ du -sh vendor/
114M	vendor/
```

**Comment by anfernee** [comment]
@dlorenc @r2d4 anything issues with this PR? Looking forward to hearing your feedback.

**Comment by frapposelli** [comment]
@gbraad this would tie in nicely with the improved vmware driver that @anfernee just moved to `github.com/machine-drivers`. 

Can you guys let us know what are the next steps to move this forward?

/cc: @dlorenc @r2d4

**Comment by dlorenc** [comment]
Looks like there's a conflict in kubeadm.go. Woul you mind rebasing to kick off a final test run?

**Comment by anfernee** [comment]
The OSX test looks suspicious. I would go grab a mac and test it. I assume it would be either red herring or small change. The rest of this PR should be ready for review. @dlorenc @r2d4 if you guys can take a look at the basic structure of this PR and give me some feedback, it would be great. thanks!

**Comment by anfernee** [comment]
I tried to run this PR on OSX, it works well. Still not sure why test fails though.

**Comment by frapposelli** [comment]
maybe a fluke? does 

/retest

work here?

**Comment by dlorenc** [comment]
I think the mac slave needed a reboot. I'll kick it off again after the machine comes back up.

File: pkg/minikube/cluster/cluster.go [hunk 1]
```diff
@@ -308,6 +296,7 @@ func MountHost(api libmachine.API, ip net.IP, path, port, mountVersion string, u
 }
 
 // GetVMHostIP gets the ip address to be used for mapping host -> VM and VM -> host
+// TODO: remove this (why not just using driver.GetIP?)
```
Comment (dlorenc) [inline pkg/minikube/cluster/cluster.go]: This is kind of the opposite - driver.GetIP is the IP address of the VM from the host, this is the IP of the host from the VM.

File: pkg/minikube/cluster/cluster_test.go [hunk 2]
```diff
@@ -67,15 +68,17 @@ func TestCreateHost(t *testing.T) {
 	}
 
 	found := false
-	for _, driver := range constants.SupportedVMDrivers {
-		if h.DriverName == driver {
+	drivers := []string{}
```
Comment (dlorenc) [inline pkg/minikube/cluster/cluster_test.go]: If you add a "String()" method to DriverDef that returns the .Name field, I think you can simplify this a bit and avoid collecting the drivers slice.

File: pkg/minikube/cluster/cluster.go (hunk 1 above)
Comment (anfernee) [reply pkg/minikube/cluster/cluster.go]: OK. I see. I am removing this comment. but i still want to see if it's possible to remove the switch for driver names sometime in future.

**Comment by anfernee** [comment]
done, thanks @dlorenc

**Comment by afbjorklund** [comment]
Is there any quick guide on how to convert a driver to this new regime ?

**Comment by praveenkumar** [comment]
@anfernee Can you also send a PR to minishift side? I think it would be good if you send instead we copy stuff and then paste.

**Comment by anfernee** [comment]
thanks @dlorenc 

@afbjorklund Good question. Not yet. but I am going to write one maybe in the drivers folder. @praveenkumar will do. but I would need to get myself familiar to minishift source. I don't mind if you guys want to copy&paste it.

**Comment by praveenkumar** [comment]
> will do. but I would need to get myself familiar to minishift source. 

@anfernee  Thanks, no hurry from our side (Take your time)

> I don't mind if you guys want to copy&paste it.

Right but I really like you to put that code instead we copy&paste :)

**Pull request merged**